### PR TITLE
Launcher: Added Launcher model with more abilities to filter

### PR DESCRIPTION
**Pull request by iLLiCiTiT**
## Brief description
Launcher has global model with single access to docs used in all widgets. Also gives ability to easily change filters for all components in launcher tool.

## Changes
- created `LauncherModel` which cares about refreshing of data (projects, assets and tasks)
    - single access to mongo database for most of launcher components
    - cares about filters for all components (filters are not added as real widgets just prepared backend for easy change of it)
- task can have `assignee` which is stored to task items in tasks widget
- assets widget in launcher has information about all children task types and all assignees on all tasks (to help filter them)

## Testing notes:
1. Launcher should work the same and maybe a little bit faster

### CODE REVIEW

**Comment by mkolar** [comment]
Task linked: [OP-2416 Launcher tool model](https://app.clickup.com/t/6658547/OP-2416)

**Comment by BigRoy** [comment]
What kind of data would the Launcher's task model or asset model have that the Asset model **must never have**. There might be data that is less important for the loader or alike but still it wouldn't get in the way?

E.g. I'd love to also allow filtering the loader assets based on assigned tasks. Or on "related to X". It seems to me that in essence the models are 99% related and where we improve one the other should too. At the very least they should inherit each other?

It will be much more work keeping functionality in sync if we separate plus there is more code to maintain.

Basically what I am asking: are we sure it's easier if we duplicate into a unique model?

Looking at this PR it only adds assignee data. I'd say that's just as worth it in the Loader or even better the Work Files browser.

**Comment by iLLiCiTiT** [comment]
Main goal of LauncherModel is to keep basic data used across multiple widgets at one place and prequery data that would be queried anyway.
For example with LauncherModel both AssetsWidget and TasksWidget do not query assets from database on their own but use LauncherModel methods to get it and listen to it's changes which gives much more clean way to decide when Asset or Task widget should change their context.

If you would want to do filters on task level you should filter on asset level too to avoid showing irrelevant assets at the same time you have to keep filtering of all widgets in sync and tell them if any has changed in a maintainable way and you have to get values that can be used for filtering (query all tasks for whole project and get them).

With all of this and threaded refresh of assets widget it was horror to make it work and read the code afterwards to understand what is happening why and understand sequence of events and avoid querying of all assets from whole project 4times on first project selection.

> At some point when something must be unique to function we can likely inherit from the other to only add slightly more data into it and still use a shared base model.

That is exactly what is happening here. `LauncherAssetModel` inherits from `AssetModel`. Changes how assets docs are passed to the model and adds 2 roles for filtering. Both roles require to query also tasks which is not case of default asset model because tasks contain a lot of data which is massive slowdown for other tools on bigger projects. At the same time tasks are used for tasks widget so tasks widget does not have to query them all the time when asset changes and at the same time you have already collected all values that can be used for filtering.

> It will be much more work keeping functionality in sync if we separate plus there is more code to maintain.

**Comment by BigRoy** [comment]
> For example with LauncherModel both AssetsWidget and TasksWidget do not query assets from database on their own but use LauncherModel methods to get it and listen to it's changes which gives much more clean way to decide when Asset or Task widget should change their context.
> 
> If you would want to do filters on task level you should filter on asset level too to avoid showing irrelevant assets at the same time you have to keep filtering of all widgets in sync and tell them if any has changed in a maintainable way and you have to get values that can be used for filtering (query all tasks for whole project and get them).

This makes sense. Thanks for the information. Should have probably spent more time investigating the whole code. 👍

**Review by BigRoy — CHANGES_REQUESTED** [review]
### Getting some errors

Getting this error when adding filter text into the filter field, going out of a project and going into a new project (the filter field is preserved) and this error is generated:
```python
Traceback (most recent call last):
  File "S:\openpype\OpenPype\openpype\tools\launcher\models.py", line 751, in filterAcceptsRow
    if self.filterAcceptsRow(child_row, source_index):
  File "S:\openpype\OpenPype\openpype\tools\launcher\models.py", line 732, in filterAcceptsRow
    if not re.search(self._name_filter, name, re.IGNORECASE):
  File "C:\Users\Roy\AppData\Local\Programs\Python\Python37\lib\re.py", line 185, in search
    return _compile(pattern, flags).search(string)
TypeError: expected string or bytes-like object
```

Getting this when trying to launch an application:
```python
Traceback (most recent call last):
  File "S:\openpype\OpenPype\openpype\tools\launcher\widgets.py", line 325, in on_clicked
    self.launcher_model.start_refresh_timer()
AttributeError: 'ActionBar' object has no attribute 'launcher_model'
```

**Comment by BigRoy** [comment]
> - task can have assignee which is stored to task items in tasks widget
> - assets widget in launcher has information about all children task types and all assignees on all tasks (to help filter them)

How would I access or test this? I can't seem to find any complex filtering other than the name field.

**Comment by iLLiCiTiT** [comment]
>  I can't seem to find any complex filtering other than the name field.

I didn't implemented widgets for the filtering yet, just prepared base to be able do it. This changes how data in launcher widgets are handled which is already a lot of changes just to have at least the same functionality.

PR adding the widgets will be probably stuck for some time as discussion about "how", "where", etc. will be raised...

**Review by BigRoy — APPROVED** [review]
> Getting this error when adding filter text into the filter field, going out of a project and going into a new project (the filter field is preserved) and this error is generated:
> ```python
> Traceback (most recent call last):
>   File "S:\openpype\OpenPype\openpype\tools\launcher\models.py", line 751, in filterAcceptsRow
>     if self.filterAcceptsRow(child_row, source_index):
>   File "S:\openpype\OpenPype\openpype\tools\launcher\models.py", line 732, in filterAcceptsRow
>     if not re.search(self._name_filter, name, re.IGNORECASE):
>   File "C:\Users\Roy\AppData\Local\Programs\Python\Python37\lib\re.py", line 185, in search
>     return _compile(pattern, flags).search(string)
> TypeError: expected string or bytes-like object
> ```

This error is still present - however had a quick test in the version before this PR. The error was already present so is **not** related to this PR.

---

> Getting this when trying to launch an application:
> ```python
> Traceback (most recent call last):
>   File "S:\openpype\OpenPype\openpype\tools\launcher\widgets.py", line 325, in on_clicked
>     self.launcher_model.start_refresh_timer()
> AttributeError: 'ActionBar' object has no attribute 'launcher_model'
> ```

This is **fixed**. I can launch applications now.
